cmd/contour/serve.go

```diff
@@ -161,32 +163,70 @@ func registerServe(app *kingpin.Application) (*kingpin.CmdClause, *serveContext)
 	return serve, ctx
 }
 
-// doServe runs the contour serve subcommand.
-func doServe(log logrus.FieldLogger, ctx *serveContext) error {
+type Serve struct {
+	group    workgroup.Group
+	log      logrus.FieldLogger
+	ctx      *serveContext
+	clients  *k8s.Clients
+	mgr      manager.Manager
+	registry *prometheus.Registry
+}
+
+func NewServe(log logrus.FieldLogger, ctx *serveContext) (*Serve, error) {
+
+	// Set up workgroup runner.
+	var group workgroup.Group
+
 	// Establish k8s core & dynamic client connections.
 	clients, err := k8s.NewClients(ctx.Config.Kubeconfig, ctx.Config.InCluster)
 	if err != nil {
-		return fmt.Errorf("failed to create Kubernetes clients: %w", err)
+		return nil, fmt.Errorf("failed to create Kubernetes clients: %w", err)
 	}
 
-	// Set up workgroup runner.
-	var g workgroup.Group
-
 	scheme, err := k8s.NewContourScheme()
 	if err != nil {
 		log.WithError(err).Fatal("unable to create scheme")
 	}
 
+	// Instantiate a controller-runtime manager.
+	mgr, err := manager.New(controller_config.GetConfigOrDie(), manager.Options{
+		Scheme: scheme,
+	})
+	if err != nil {
+		log.WithError(err).Fatal("unable to set up controller manager")
+	}
+
+	// Set up Prometheus registry and register base metrics.
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
+	registry.MustRegister(collectors.NewGoCollector())
+
+	return &Serve{
+		group:    group,
+		log:      log,
+		ctx:      ctx,
+		clients:  clients,
+		mgr:      mgr,
+		registry: registry,
+	}, nil
+}
+
+// doServe runs the contour serve subcommand.
+func (s *Serve) doServe() error {
+
+	// Convert the Contour ServeContext into a ContourConfiguration
+	contourConfiguration := s.ctx.convertToContourConfigurationSpec()
```

Comment:
Only need to do this if `s.ctx.contourConfigurationName` is empty, right?

Reply:
Related, on L237 we're declaring a new `contourConfiguration` var inside the if statement and populating it with the CRD contents, but we should be populating the var declared here on L218.

Reply:
the initial point of this was to only covert the servecontext to the crd spec. The next one was to then actually implement and use it, but can do that all here in this PR if you want.

Reply:
I just updated to use it the config if passed in, just another line of code.

Reply:
Ah, well if it's only one more line to use the CRD when present may as well :)